docs/strings.md

```diff
@@ -0,0 +1,464 @@
+By pairing these, the caller can access the character data in the string's native encoding, possibly without the need to copy, and operations on two strings that happen to be in the same encoding can be significantly faster.
+
+This design is very flexible and allows the string to use whatever internal storage it wants (for example, ASCII / UTF-8 storage) while still allowing operations on a range of characters to do only one dynamic call.
+It does not; however, allow for compile-time specialisation when the concrete type of the strings is known at compile time, which eliminates a lot of microoptimisation opportunities.
+
+This basic pattern was extended for collections in the `NSFastEnumeration` protocol and ICU's `UText` abstraction provides a similar model.
+The optimization goals for `UText` are slightly different: all of the operations on `UText` work with UTF-16 data and so the responsibility for converting to UTF-16 is delegated to individual string objects.
+In `UText`, the generic text object has three fields containing a pointer to a buffer, the character at the start of the buffer, and the length of the buffer, along with a virtual function that can fill the buffer.
+The generic `UText` allocator also provides an option to allocate space at the end of the object that instances can use.
+The fill function can either set the data field to point to its internal storage, copy into the space at the end, or allocate additional memory to store a temporary buffer.
+
+It is worth noting that `NSString` / `NSMutableString` and `UText` provide sufficiently similar abstractions that it is fairly trivial (under 500 lines of code) to create [bidirectional bridges between the two](https://github.com/gnustep/libs-base/blob/master/Source/GSICUString.m).
+
+String views
+------------
+
+C++, C#, and other languages have recently introduced the idea of a view within a string, similar to a Module-2 slice.
+A string view gives a range within a string but does not duplicate the underlying storage.
+If a string view is mutable then changes in it will be reflected into the underlying storage.
+
+String views in Verona can use the region abstraction to ensure correct lifetime management.
+A string view on a mutable string will be in the same region as the underlying string and it is the responsibility of that region's memory management to ensure that the region does not outlive the string view.
+Alternatively, string views may be stack types that are bound to a specific region.
+
+String views that refer to immutable strings do not introduce memory management difficulties in Verona because immutable objects are reference counted.
+Even if the string view is mutable, it can be in any region without causing problems.
```

Comment:
I think we need some extra words here.  An `imm & StringView` can only refer to an `imm & String` because an `imm` type can't refer to a mutable object, but you may want to have a read-only view of a portion of a string.  A read-only view of a mutable string is actually pretty problematic because someone else can mutate the string out from under you.

You could have an `iso` (or `mut`) `StringView` that is a mutable view (i.e. you can change the bounds) of an immutable string (i.e. the string is `String & imm`).  That can be useful if you want to have a moving window over a string, where something captures the `StringView` and you update it.  I don't really like this though, I'd much rather that `StringView`s were treated as value types and can modify a string if the underlying string is mutable.

Reply:
Right, so here you refer to the string views that are mutable (offset, range) on an immutable string (contents).

Suppose I create a function that takes in a mutable `StringView` to change the underlying data. But if I pass a mutable view on an immutable string, I'll trigger an error trying to write to it (if I don't catch it at compile time).

How do I differentiate, in the syntax of the argument's type, between a "mutable string view" over a "mutable string" from one over an "immutable string"?

Reply:
Right, I think this is an open question and one I'd be interested in feedback on.  My guess is a mutable `StringView` should always refer to mutable data, my use case for a mutable view of an immutable string feels contrived.